src/garbage_collector.rs

```diff
@@ -0,0 +1,253 @@
+                .expect("failed to stop main compactor thread");
+        }
+    }
+}
+
+struct GarbageCollectorOrchestrator {
+    manifest_store: Arc<ManifestStore>,
+    table_store: Arc<TableStore>,
+    options: GarbageCollectorOptions,
+    external_rx: crossbeam_channel::Receiver<GarbageCollectorMessage>,
+    db_stats: Arc<DbStats>,
+}
+
+impl GarbageCollectorOrchestrator {
+    /// Collect garbage from the manifest store. This will delete any manifests
+    /// that are older than the minimum age specified in the options.
+    async fn collect_garbage_manifests(&self) -> Result<(), SlateDBError> {
+        let min_age = self
+            .options
+            .manifest_options
+            .map_or(DEFAULT_MIN_AGE, |opts| opts.min_age);
+        let mut manifest_metadata_list = self.manifest_store.list_manifests(..).await?;
+
+        // Remove the last element so we never delete the latest manifest
+        manifest_metadata_list.pop();
+
+        // TODO Should exclude snapshotted manifests when we implement snapshots
+
+        // Delete manifests older than min_age
+        for manifest_metadata in manifest_metadata_list {
+            let min_age = chrono::Duration::from_std(min_age).expect("invalid duration");
+
+            if Utc::now().signed_duration_since(manifest_metadata.last_modified) > min_age {
+                if let Err(e) = self
+                    .manifest_store
+                    .delete_manifest(manifest_metadata.id)
+                    .await
+                {
+                    error!("Error deleting manifest: {}", e);
+                } else {
+                    self.db_stats.gc_manifest_count.inc();
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Collect garbage from the WAL SSTs. This will delete any WAL SSTs that are
+    /// older than the minimum age specified in the options and are also older than
+    /// the last compacted WAL SST.
+    async fn collect_garbage_wal_ssts(&self) -> Result<(), SlateDBError> {
+        let last_compacted_wal_sst_id = self
+            .manifest_store
+            // Get the latest manifest so we can get the last compacted id
+            .read_latest_manifest()
+            .await?
+            .ok_or_else(|| SlateDBError::ManifestMissing)?
+            // read_latest_manifest returns (id, manifest) but we only care about the manifest
+            .1
+            .core
+            .last_compacted_wal_sst_id;
+        let min_age = self
+            .options
+            .wal_options
+            .map_or(DEFAULT_MIN_AGE, |opts| opts.min_age);
+        let sst_ids_to_delete = self
+            .table_store
+            .list_wal_ssts(..last_compacted_wal_sst_id)
+            .await?
+            .into_iter()
+            .filter(|wal_sst| {
+                let min_age = chrono::Duration::from_std(min_age).expect("invalid duration");
+                Utc::now().signed_duration_since(wal_sst.last_modified) > min_age
```

Comment:
nit: `Utc:now()` used for every file would not be the same if there are a lot of files to delete. Using a single timestamp for one round of delete would be more consistent.